### PR TITLE
COOPR-496 add links to the cluster template that can point to

### PR DESCRIPTION
**Pull request by albertshau**
various useful service UIs. Services and cluster templates have
a links field that consists of an array of links, where a link
is a label and a url. The url can contain macros. When a cluster
is fetched from the rest api, links in the cluster template are
expanded and placed at the top level of the returned object.
In addition, for every node in the cluster, the links for all
services on the node are taken, expanded, and combined into a list
of links for the node.

### CODE REVIEW

**Comment by wolf31o2** [comment]
Could you provide an example of what "links" may look like for the `hadoop-hdfs-namenode` service?

Also, if no services had any links, but there were links in a clustertemplate, only those links would be returned at the top-level and none would be returned per-node, correct?

**Comment by albertshau** [comment]
yup so you might have for a service:

```
{
    "name":"hadoop-hdfs-namenode",
    "links":[
        {
            "label": "HDFS Namenode",
            "url": "http://%ip.access_v4.self%:50070"
        }
    ],
    ...
}
```

and then for a template you might have:

```
{
    "name":"hadoop-distributed",
    "links":[
        {
            "label": "HDFS Namenode",
            "url": "http://%ip.access_v4.service.hadoop-hdfs-namenode%:50070"
        },
        {
            "label": "HDFS Resource Manager",
            "url": "http://%ip.access_v4.service.hadoop-yarn-resourcemanager%:8088"
        },
        {
            "label": "HBase Master",
            "url": "http://%ip.access_v4.service.hbase-master%:60010"
        },
    ],
    ...
}
```

**Comment by albertshau** [comment]
and then these would get expanded on a `GET /clusters/<id>` to:

```
{
    "name":"my-cluster-name",
    "links": [
        {
            "label": "HDFS Namenode",
            "url": "http://1.2.3.4:50070"
        },
        {
            "label": "HDFS Resource Manager",
            "url": "http://1.2.3.4:8088"
        },
        {
            "label": "HBase Master",
            "url": "http://1.2.3.4:60010"
        }
    ],
    ...
    "nodes": [
        {
            // assume this is the namenode
            "id": ...,
            "clusterId": ...,
            "links": [
                 {
                    "label": "HDFS Namenode",
                    "url": "http://1.2.3.4:50070"
                },
                {
                    "label": "HDFS Resource Manager",
                    "url": "http://1.2.3.4:8088"
                },
                {
                    "label": "HBase Master",
                    "url": "http://1.2.3.4:60010"
                }
            ],
            ...
        },
        ...    
    ]
}
```

**Comment by wolf31o2** [comment]
Cool. Looks as I expected. Thanks.

File: coopr-server/src/main/java/co/cask/coopr/codec/json/current/ServiceCodec.java [hunk 1]
```diff
@@ -65,8 +68,9 @@ public Service deserialize(JsonElement json, Type type, JsonDeserializationConte
     Map<ProvisionerAction, ServiceAction> actions = Collections.emptyMap();
     if (provisioner != null) {
       actions = context.deserialize(provisioner.get("actions"),
-                                   new TypeToken<Map<ProvisionerAction, ServiceAction>>(){}.getType());
+                                    new TypeToken<Map<ProvisionerAction, ServiceAction>>() {}.getType());
```
Comment (poornachandra) [inline coopr-server/src/main/java/co/cask/coopr/codec/json/current/ServiceCodec.java]: new TypeToken<Map<ProvisionerAction, ServiceAction>>() {}.getType() can be a constant.

File: coopr-server/src/main/java/co/cask/coopr/codec/json/current/ServiceCodec.java [hunk 2]
```diff
@@ -65,8 +68,9 @@ public Service deserialize(JsonElement json, Type type, JsonDeserializationConte
...
     }
+    Set<Link> links = context.deserialize(jsonObj.get("links"), new TypeToken<Set<Link>>() {}.getType());
```
Comment (poornachandra) [inline coopr-server/src/main/java/co/cask/coopr/codec/json/current/ServiceCodec.java]: new TypeToken<Set<Link>>() {}.getType() can be a constant.

File: coopr-server/src/main/java/co/cask/coopr/cluster/Node.java [hunk 3]
```diff
@@ -31,14 +35,14 @@
...
-public final class Node implements Comparable<Node> {
-  private static final String IPADDRESS_KEY = "ipaddress";
+public class Node implements Comparable<Node> {
   private final String id;
   private final String clusterId;
   private final Set<Service> services;
   private final NodeProperties properties;
   private final List<Action> actions;
   private final JsonObject provisionerResults;
+  private List<Link> links;
```
Comment (poornachandra) [inline coopr-server/src/main/java/co/cask/coopr/cluster/Node.java:45]: links is not exposed by any public method. What will links be used for?

File: coopr-server/src/main/java/co/cask/coopr/codec/json/current/ClusterTemplateCodec.java [hunk 4]
```diff
@@ -62,6 +66,7 @@ public ClusterTemplate deserialize(JsonElement json, Type type, JsonDeserializat
       .setCompatibilities(context.<Compatibilities>deserialize(jsonObj.get("compatibility"), Compatibilities.class))
       .setConstraints(context.<Constraints>deserialize(jsonObj.get("constraints"), Constraints.class))
       .setAdministration(context.<Administration>deserialize(jsonObj.get("administration"), Administration.class))
+      .setLinks(context.<Set<Link>>deserialize(jsonObj.get("links"), new TypeToken<Set<Link>>() {}.getType()))
```
Comment (poornachandra) [inline coopr-server/src/main/java/co/cask/coopr/codec/json/current/ClusterTemplateCodec.java]: new TypeToken<Set<Link>>() {}.getType() can be a constant.

**Comment by poornachandra** [comment]
Minor comments, otherwise LGTM :+1:

File: coopr-server/src/main/java/co/cask/coopr/cluster/Node.java (hunk 3 above)
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/Node.java:45]: will add a getter.  Not currently used, but may be used by some cluster callback for example.

**Pull request merged**